### PR TITLE
feat(storage): add storage support to application resource

### PR DESCRIPTION
**Pull request by anvial**
## Description

This PR adds storage support for application resources.

## Type of change

- Change existing resource

## QA steps

```tf
terraform {
  required_providers {
    juju = {
      source  = "juju/juju"
      version = "0.12.0"
    }
  }
}
provider "juju" {}

resource "juju_model" "storage-model" {
  name = "storage-model"
}

resource "juju_application" "postgresql" {
  name  = "postgresql"
  model = juju_model.storage-model.name

  charm {
    name     = "postgresql"
    channel  = "latest/stable"
  }

  storage_directives = { pgdata = "8M" }

  units = 1
}
...
```

### CODE REVIEW

File: internal/provider/resource_application.go [hunk 1]
```diff
@@ -122,6 +129,92 @@ func (r *applicationResource) Configure(ctx context.Context, req resource.Config
 	r.subCtx = tflog.NewSubsystem(ctx, LogResourceApplication)
 }
 
+func storageSetRequiresReplace(ctx context.Context, req planmodifier.SetRequest, resp *setplanmodifier.RequiresReplaceIfFuncResponse) {
```
Comment (hmlanigan) [inline internal/provider/resource_application.go:132]: This belongs in a new file, `modifer_storageset.go`, as the validators are.

File: internal/provider/resource_application.go [hunk 2]
```diff
@@ -122,6 +129,92 @@ func (r *applicationResource) Configure(ctx context.Context, req resource.Config
...
+					Size:  parsedStorageSize,
+					Pool:  storagePool,
+					Count: uint64(storageCount),
+				}
+			}
+		}
+	}
+
+	stateSet := make(map[string]jujustorage.Constraints)
+	if !req.StateValue.IsNull() {
```
Comment (hmlanigan) [inline internal/provider/resource_application.go:165]: Investigate whether you need to check req.StateValue.IsUnknown() also

File: internal/provider/resource_application.go (hunk 1 above)
Comment (hmlanigan) [inline internal/provider/resource_application.go:132]: Needs function header comments.

File: internal/provider/resource_application.go [hunk 3]
```diff
@@ -373,6 +515,18 @@ func (n nestedEndpointBinding) transformToStringTuple() (string, string) {
...
+// nestedStorage represents the single element of the storage SetNestedAttribute
+type nestedStorage struct {
+	Label types.String `tfsdk:"label"`
+	Size  types.String `tfsdk:"size"`
+	Pool  types.String `tfsdk:"pool"`
+	Count types.Int64  `tfsdk:"count"`
+}
+
+func (n nestedStorage) transformToString() string {
+	return fmt.Sprintf("%s=%s,%s,%d", n.Label.ValueString(), n.Size.ValueString(), n.Pool.ValueString(), n.Count.ValueInt64())
```
Comment (hmlanigan) [inline internal/provider/resource_application.go:527]: This is a good place to use String() rather than ValueString(). It'll do the quoting for you. The alternative is to use `%q` rather than `%s`.

File: internal/provider/resource_application.go [hunk 4]
```diff
@@ -577,6 +771,7 @@ func (r *applicationResource) Read(ctx context.Context, req resource.ReadRequest
 		return
 	}
 	r.trace(fmt.Sprintf("read application resource %q", appName))
+	r.trace("ReadApplication response", map[string]interface{}{"response": response})
```
Comment (hmlanigan) [inline internal/provider/resource_application.go:774]: combine, no need for 2.

File: internal/provider/resource_application.go [hunk 5]
```diff
@@ -636,6 +831,22 @@ func (r *applicationResource) Read(ctx context.Context, req resource.ReadRequest
...
+
+		nestedStorageSlice = append(nestedStorageSlice, nestedStorage{
+			Label: types.StringValue(name),
+			Size:  types.StringValue(humanizedSize),
+			Pool:  types.StringValue(storage.Pool),
+			Count: types.Int64Value(int64(storage.Count)),
+		})
+	}
+	storageType := req.State.Schema.GetAttributes()[StorageKey].(schema.SetNestedAttribute).NestedObject.Type()
+	state.Storage, dErr = types.SetValueFrom(ctx, storageType, nestedStorageSlice)
```
Comment (hmlanigan) [inline internal/provider/resource_application.go:848]: Missing a check of dErr.HasError() after this line.

File: internal/provider/resource_application.go [hunk 6]
```diff
@@ -875,6 +1086,61 @@ func (r *applicationResource) Update(ctx context.Context, req resource.UpdateReq
 		updateApplicationInput.EndpointBindings = endpointBindings
 	}
 
+	// Check if we have new storages in plan that not existed in the state, and add their constraints to the
```
Comment (hmlanigan) [inline internal/provider/resource_application.go:1089]: s/storages/storage/

Comments should be full sentence with a period.

File: internal/provider/resource_application.go [hunk 7]
```diff
@@ -875,6 +1086,61 @@ func (r *applicationResource) Update(ctx context.Context, req resource.UpdateReq
...
+			return
+		}
+
+		// Map of storage name to storage constraints
+		planStorageMap := make(map[string]jujustorage.Constraints)
+		for _, storage := range planStorageSlice {
+			// Validate storage size
+			parsedSize, err := utils.ParseSize(storage.Size.ValueString())
+			if err != nil {
+				return
```
Comment (hmlanigan) [inline internal/provider/resource_application.go:1108]: Before returning you need to call resp.Diagnostics.AddError() or this won't be caught by terraform.

File: internal/provider/resource_application.go [hunk 8]
```diff
@@ -875,6 +1086,61 @@ func (r *applicationResource) Update(ctx context.Context, req resource.UpdateReq
...
+				Pool:  storage.Pool.ValueString(),
+				Count: uint64(storage.Count.ValueInt64()),
+			}
+		}
+		stateStorageMap := make(map[string]jujustorage.Constraints)
+		for _, storage := range stateStorageSlice {
+			// Validate storage size
+			parsedSize, err := utils.ParseSize(storage.Size.ValueString())
+			if err != nil {
+				return
```
Comment (hmlanigan) [inline internal/provider/resource_application.go:1121]: resp.Diagnostics.AddError()

File: internal/provider/resource_application.go [hunk 9]
```diff
@@ -875,6 +1086,61 @@ func (r *applicationResource) Update(ctx context.Context, req resource.UpdateReq
 		updateApplicationInput.EndpointBindings = endpointBindings
 	}
 
+	// Check if we have new storages in plan that not existed in the state, and add their constraints to the
+	// update application input
+	if !plan.Storage.Equal(state.Storage) {
```
Comment (hmlanigan) [inline internal/provider/resource_application.go:1091]: Please move to a new method. Update() is becoming too long and unwheldy.

**Review by hmlanigan — COMMENTED** [review]
Here is my initial review. I haven't done QA yet. There is a conflict and perhaps some tests to fix as well.

File: internal/juju/applications.go [hunk 10]
```diff
@@ -816,9 +893,17 @@ func (c applicationsClient) ReadApplication(input *ReadApplicationInput) (*ReadA
 	appInfo := apps[0].Result
 
 	// TODO: Investigate why we're getting the full status here when
-	// application status is needed.
-	status, err := clientAPIClient.Status(nil)
+	// application status is needed. include storage.
```
[the remark this answers is not in the document]
Comment (anvial) [reply internal/juju/applications.go:896]: rewrote comment

File: internal/provider/resource_application.go (hunk 1 above)
Comment (anvial) [reply internal/provider/resource_application.go:132]: Not sure that I know where to put this file.

File: internal/provider/resource_application.go (hunk 2 above)
Comment (anvial) [reply internal/provider/resource_application.go:167]: it does not need to happen because we use this function in Update.

File: internal/provider/resource_integration.go [hunk 11]
```diff
@@ -222,6 +222,10 @@ func (r *integrationResource) Create(ctx context.Context, req resource.CreateReq
 
 	appsType := req.Plan.Schema.GetBlocks()["application"].(schema.SetNestedBlock).NestedObject.Type()
 	parsedApps, errDiag := types.SetValueFrom(ctx, appsType, parsedApplications)
+	if errDiag.HasError() {
+		resp.Diagnostics.Append(errDiag...)
+		return
+	}
```
Comment (hmlanigan) [inline internal/provider/resource_integration.go:228]: Why did you add this? The errDiag is being checked, only after being added to the response.Diagnostics.

File: internal/provider/resource_application_test.go [hunk 12]
```diff
@@ -507,6 +507,33 @@ func TestAcc_ResourceApplication_UpdateEndpointBindings(t *testing.T) {
 	})
 }
 
+func TestAcc_ResourceApplication_Storage(t *testing.T) {
```
Comment (hmlanigan) [inline internal/provider/resource_application_test.go:510]: This Test shouldn't run on microk8s, it doesn't support storage.

File: internal/provider/resource_application_test.go [hunk 13]
```diff
@@ -580,17 +607,21 @@ resource "juju_application" "{{.AppName}}" {
...
+  }]
+  {{- end }}
 
-  {{ if ne .ResourceParamName "" }}
+  {{- if ne .ResourceParamName "" }}
   resources = {
     {{.ResourceParamName}} = {{.ResourceParamRevision}}
   }
-  {{ end }}
+  {{- end }}
```
Comment (hmlanigan) [inline internal/provider/resource_application_test.go:624]: This change is having unintended consequences in other tests I think, I see new failures in TestAcc_ResourceOffer and TestAcc_ResourceIntegrationWithMultipleConsumers.

File: internal/provider/resource_integration.go (hunk 11 above)
Comment (anvial) [reply internal/provider/resource_integration.go:228]: Unintentionally, I think I was fixing issues with `golangci-lint`.
Let me revert the changes.

File: internal/provider/resource_application_test.go (hunk 12 above)
Comment (anvial) [reply internal/provider/resource_application_test.go:510]: fixed

File: internal/provider/resource_application_test.go (hunk 13 above)
Comment (anvial) [reply internal/provider/resource_application_test.go:624]: That strange, because these changes only remove empty lines from the result plan.
But OK I can revert them.

**Review by hmlanigan — COMMENTED** [review]
I added 2 commits, one to enable storage for juju controllers not using DeployFromRepository and to remove the Private data usage. 

Once the rest of these smaller comments are addressed, we should be good to land.

File: examples/resources/juju_application/resource.tf [hunk 14]
```diff
@@ -17,7 +17,7 @@ resource "juju_application" "this" {
   }
 }
 
-resource "juju_application" "placement_example" {
+resource "juju_application" "placement_and_storage_example" {
```
[the remark this answers is not in the document]
Comment (anvial) [reply examples/resources/juju_application/resource.tf:20]: done

File: internal/juju/applications.go [hunk 15]
```diff
@@ -364,6 +384,7 @@ func (c applicationsClient) deployFromRepository(applicationAPIClient *apiapplic
 		Revision:         &transformedInput.charmRevision,
 		Trust:            transformedInput.trust,
 		Resources:        resources,
+		Storage:          transformedInput.storage,
```
[the remark this answers is not in the document]
Comment (anvial) [reply internal/juju/applications.go:387]: thx!

That's was my miss. I wish to change at the beginning but the just forgot to do it.

File: internal/juju/applications.go [hunk 16]
```diff
@@ -318,6 +335,9 @@ func (c applicationsClient) CreateApplication(ctx context.Context, input *Create
 		return nil, err
 	}
 
+	// print transformedInput.storage
+	c.Tracef("transformedInput.storage", map[string]interface{}{"transformedInput.storage": transformedInput.storage})
```
[the remark this answers is not in the document]
Comment (anvial) [reply internal/juju/applications.go:339]: nope, I don't think so. Thx

**Review by hmlanigan — APPROVED** [review]
LGTM! Thank you!
